src/pepr/operator/reconcilers/package-reconciler.ts

```diff
@@ -107,3 +110,51 @@ export async function packageReconciler(pkg: UDSPackage) {
     void handleFailure(err, pkg);
   }
 }
+
+/**
+ * The finalizer is called when an update with a deletion timestamp happens.
+ * On completion the finalizer is removed from the Package CR.
+ * This function removes any SSO/Authservice clients and ensures that Istio Injection is restored to the original state.
+ *
+ * @param pkg the package to finalize
+ */
+export async function packageFinalizer(pkg: UDSPackage) {
+  log.debug(`Processing removal of package ${pkg.metadata?.namespace}/${pkg.metadata?.name}`);
+
+  // In order to avoid triggering a second call of this finalizer, we just write events for each removal piece
+  // This could be switched to updateStatus once https://github.com/defenseunicorns/pepr/issues/1316 is resolved
+  // await updateStatus(pkg, { phase: Phase.Removing });
+
+  try {
+    await writeEvent(pkg, {
+      message: `Restoring original istio injection status on namespace`,
+      reason: "RemovalInProgress",
+      type: "Normal",
+    });
+    // Cleanup the namespace
+    await cleanupNamespace(pkg);
+  } catch (e) {
+    await writeEvent(pkg, {
+      message: `Restoration of istio injection status failed, ${e.message}`,
+      reason: "RemovalFailed",
+    });
+  }
```

Comment:
Obviously catching these and just tossing an event is non-ideal but in the current state of `Finalize` we don't have any options to fail the function here without deleting the `finalizer` from the CR ([ref](https://github.com/defenseunicorns/pepr/blob/63aebabb135099d523fe08dc220fab84027e1dc5/src/lib/watch-processor.ts#L113-L115)). This is also still a slight improvement over the previous state since we are able to leave some breadcrumb about the failure, whereas previously the Package CR instantly deleted with no events indicating progress or failure.